extensions/csv/src/main/java/io/deephaven/csv/CsvTools.java

```diff
@@ -913,35 +885,321 @@ private static void writeCsvContentsSeq(
+                @NotNull final Writer writer,
+                @NotNull final RowSequence rows,
+                @Nullable final BiConsumer<Long, Long> progress) throws IOException {
+            final long totalSize = rows.size();
+            long rowsWritten = 0;
+            try (final RowSequence.Iterator rowsIterator = rows.getRowSequenceIterator()) {
+                while (rowsIterator.hasMore()) {
+                    final RowSequence sliceRows = rowsIterator.getNextRowSequenceWithLength(CHUNK_CAPACITY);
+                    final int sliceSize = sliceRows.intSize();
+                    for (final ColumnFormatter<?> columnFormatter : columnFormatters) {
+                        columnFormatter.fill(sliceRows);
+                    }
+                    for (int sri = 0; sri < sliceSize; ++sri) {
+                        for (int ci = 0; ci < columnFormatters.length; ++ci) {
+                            if (ci == 0) {
+                                // Writing our line separator at the beginning, rather than the end, seems like an
+                                // unusual choice, but the code has behaved this way for some time. Maybe related to
+                                // the multi-table writing functionality and `tableSeparator`?
+                                writer.write(lineSeparator);
+                            } else {
+                                writer.write(separator);
+                            }
+                            columnFormatters[ci].write(writer, sri);
+                        }
                     }
-                    final Object o = cols[ci].get(rowKey);
-                    if (o instanceof String) {
-                        out.write("" + separatorCsvEscape((String) o, separatorStr));
-                    } else if (o instanceof Instant) {
-                        final ZonedDateTime zdt = ZonedDateTime.ofInstant((Instant) o, timeZone);
-                        out.write(separatorCsvEscape(
-                                zdt.toLocalDateTime().toString() + zdt.getOffset().toString(), separatorStr));
-                    } else {
-                        out.write(nullsAsEmpty
-                                ? separatorCsvEscape(o == null ? "" : o.toString(), separatorStr)
-                                : separatorCsvEscape(TableTools.nullToNullString(o), separatorStr));
+                    if (sharedContext != null) {
+                        sharedContext.reset();
                     }
+                    if (progress != null) {
+                        progress.accept(rowsWritten += sliceSize, totalSize);
+                    }
+                }
+            }
+        }
+
+        @Override
+        public void close() {
+            SafeCloseable.closeAll(Stream.concat(Stream.of(sharedContext), Arrays.stream(columnFormatters)));
+        }
+
+        private ColumnFormatter<?> makeColumnFormatter(@NotNull final ColumnSource<?> source) {
```

Comment:
No boolean support?

Reply:
Booleans are handled as boxed objects, to allow nullability. There are no Deephaven ColumnSources that would return a `BooleanChunk`.